### PR TITLE
Flags backend

### LINKED ISSUE #20: Command line flags backend
We should have a backend that uses the [`flag`](https://golang.org/pkg/flag/) package to load conf from the command line.

### PR DESCRIPTION
**Pull request by asdine**
This PR adds the support of the flags backend.
A new interface has been added `StructLoader` which is used to indicate that the backend wants to use the **parsed** struct informations directly and manipulate it.
The `ValueUnmarshaler` struct has been removed and replaced by a simpler `Unmarshaler` interface that can be used directly by json and yaml.

Fixes #20

### CODE REVIEW

File: backend/file/file.go [hunk 1]
```diff
@@ -7,20 +7,18 @@ import (
...
 type Backend struct {
-	path        string
-	unmarshaler backend.ValueUnmarshaler
-	name        string
+	path string
+	name string
 }
 
 // NewBackend creates a configuration loader that loads from a file.
-// The content will get decoded based on the file extension and cached in the backend.
```
Comment (YassineE) [inline backend/file/file.go]: Why did you remove the cache "feature" ?

Comment (asdine) [reply backend/file/file.go]: The interface is not the same anymore. `ValueUnmarshaler` -> `Unmarshaler` meaning the whole file is unmarshaled and not single values which required using a cache (otherwhile for each field we would have to reopen the file)

File: backend/flags/flags.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			v := flag.String(f.Key, f.Default.String(), "")
+			defer func() {
+				f.Value.SetString(*v)
+			}()
+		default:
+			flag.Var(&flagValue{f}, f.Key, "")
+		}
+	}
+
+	b.parseFlagFn()
```
Comment (YassineE) [inline backend/flags/flags.go]: Why do you need to store the `flag.Parse` function into the field ? 
Why don't call directly `flag.Parse()`

Comment (asdine) [reply backend/flags/flags.go]: It was a left-over of a previous technique i used to test the flags. But since i switched to calling the test binary itself it's not needed anymore :+1:

File: backend/flags/flags_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+			break
+		}
+		args = args[1:]
+	}
+	if len(args) == 0 {
+		fmt.Fprintf(os.Stderr, "No args\n")
+		os.Exit(2)
+	}
+
+	os.Args = append(os.Args[:1], args...)
```
Comment (YassineE) [inline backend/flags/flags_test.go]: What is the purpose of this line ?

Comment (asdine) [reply backend/flags/flags_test.go]: `flag.Parse` parses `os.Args`. The binary used to run this function is used with several arguments but the way the arguments are passed is not what we want Confita to parse.

Here is the command produced by the test "Use Defaults":

```
/tmp/go-build500277604/b001/flags.test -test.run=TestHelperProcess -- -a=hello -b=true -c=10s -d=-100
```

You can notice that between the `-test.run` flag and the `-a -b` flags there is `--`

This is to indicate the test binary "don't parse anything after `--`".
Without that we would have this error:

```sh
$ /tmp/go-build500277604/b001/flags.test -test.run=TestHelperProcess -a=hello -b=true -c=10s -d=-100

flag provided but not defined: -a
Usage of /tmp/go-build605850438/b001/flags.test:
  -httptest.serve string
    	if non-empty, httptest.NewServer serves on this address and blocks
  -test.bench regexp
    	run only benchmarks matching regexp
...
```

With the `--` flags we are basically *serializing* the parameters

Once the binary is called, the `TestHelperProcess` is called and its os.Args slice contains this:

`[/tmp/go-build189961190/b001/flags.test -test.run=TestHelperProcess -- -a=hello -b=true -c=10s -d=-100]`

`flag.Parse` is used by our backend to parse `os.Args`, but the content above doesn't represent a real world example like this one 

```sh
myprog -a=hello -b=true -c=10s -d=-100
```

We need to transform the content of os.Args to do that, that's what is done by the loop and the append. After that we have this content in os.Args:

```
[/tmp/go-build611018318/b001/flags.test -a-def=bye -b-def=false -c-def=15s -d-def=-200]
```

We can then call the backend and ask it to parse.

Comment (YassineE) [reply backend/flags/flags_test.go]: Thanks for this deep explanation.
